File: src/main/java/de/hinundhergestellt/jhuh/service/ready2order/ArtooMappedProduct.java
package de.hinundhergestellt.jhuh.service.ready2order;

import java.util.List;

public abstract class ArtooMappedProduct {

    private final List<ArtooMappedVariation> variations;

    protected ArtooMappedProduct(List<ArtooMappedVariation> variations) {
        this.variations = variations;
    }

    public abstract String getName();

    public List<ArtooMappedVariation> getVariations() {
        return variations;
    }

    public boolean isReadyForSync() {
        return variations.stream().allMatch(it -> it.getBarcode().isPresent());
    }
}

File: src/main/java/de/hinundhergestellt/jhuh/service/ready2order/ArtooMappedProduct.kt
package de.hinundhergestellt.jhuh.service.ready2order

import de.hinundhergestellt.jhuh.vendors.ready2order.ArtooProduct

class ArtooMappedProduct internal constructor(
    private val product: ArtooProduct,
    variations: List<ArtooMappedVariation>
) {
    val variations = variations.ifEmpty { listOf(ArtooMappedVariation(null, product)) }

    val id by product::id
    val name by product::name
    val description by product::description
    val price by product::price
    val hasOnlyDefaultVariant = variations.isEmpty()

    val barcodes
        get() = variations.mapNotNull { it.barcode }

    fun findVariationByBarcode(barcode: String) =
        variations.firstOrNull { it.barcode == barcode }

    override fun toString(): String {
        return "ArtooMappedProduct(id=$id, name='$name', description='$description')"
    }
}
